File: 128. Longest Consecutive Sequence.cpp
class Solution {
public:
    // sort: O(NlogN)time, O(1)space, 3.93%
    int longestConsecutive(vector<int>& nums) {
        if(nums.empty()) return 0;
        sort(nums.begin(), nums.end());
        int i = 0, j = 1, ret = 1;
        for(j; j < nums.size(); j++){
            if(nums[j] > nums[j - 1] + 1){ // bug: "(nums[j] != nums[j - 1] + 1", duplicates
                ret = max(ret, nums[j - 1] - nums[i] + 1);
                i = j;
            }
        }
        return max(ret, nums[j - 1] - nums[i] + 1);
    }

    // unordered_set: O(1) for insertion, retrieval and deletion
    // O(N)time & space, 81.59%, ok
    int longestConsecutive(vector<int>& nums) {
        if(nums.empty()) return 0;
        int ret = 0;
        unordered_set<int> unique;
        for(int i = 0; i < nums.size(); i++)
           unique.insert(nums[i]);
        for(int i = 0; i < nums.size(); i++){
            if(unique.empty()) break;
            if(unique.find(nums[i]) != unique.end()){  // find each consecutive truck
                int head = nums[i] - 1, tail = nums[i] + 1;
                unique.erase(nums[i]);
                while(unique.find(head) != unique.end())
                    unique.erase(head--);
                while(unique.find(tail) != unique.end())
                    unique.erase(tail++);
                ret = max(ret, tail - head - 1);
            }
        }
        return ret;
    }
};

File: 128. Longest Consecutive Sequence.py
import heapq
class Solution(object):
    
    # # sort: O(NlogN) time, O(1) space
    # def longestConsecutive(self, nums):
    #     """
    #     :type nums: List[int]
    #     :rtype: int
    #     """
    #     if nums == []: 
    #         return 0
    #     ret, cnt = 0, 1
    #     nums.sort()
    #     for i in xrange(1, len(nums)):
    #         if nums[i] == nums[i - 1]:
    #             continue
    #         elif nums[i] == nums[i - 1] + 1:
    #             cnt += 1
    #         else:
    #             ret = max(ret, cnt)
    #             cnt = 1
    #     return max(ret, cnt)
    
    # # heap: O(NlogN) time, O(1) space
    # def longestConsecutive(self, nums):
    #     if nums == []: return 0
    #     h = []
    #     for num in nums:
    #         heapq.heappush(h, num)
    #     ret, cnt, prev = 0, 1, None
    #     while h:
    #         curr = heapq.heappop(h)
    #         if prev is not None and curr == prev + 1:
    #             cnt += 1
    #         elif prev is not None and curr == prev:
    #             continue
    #         else:
    #             ret = max(ret, cnt)
    #             cnt = 1
    #         prev = curr
    #     return max(ret, cnt)
    
        # hashmap: O(N) time
    def longestConsecutive(self, nums):
        length, ret = {}, 0
        for num in nums:
            if num not in length:
                left = length.get(num - 1, 0)
                right = length.get(num + 1, 0)
                total = left + right + 1
                length[num] = length[num - left] = length[num + right] = total
                ret = max(ret, total)
        return ret
       
        
        
        